File: week_6/sentimental-credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int main(void)
{
    bool through_luhn;
    int get_length(long num);
    bool luhn(int length, long credit);

    // get the number of credit card
    long credit = get_long("Number: ");

    // calculate the length of the credit card
    int length = get_length(credit);

    //determine the credit card number is whether comply with luhn's Luhn’s Algorithm
    through_luhn = luhn(length, credit);

    // determine the category of the card
    if (through_luhn)
    {
        if ((length == 13 || length  == 16) && (int)(credit / pow(10, length - 1)) == 4)
        {
            printf("VISA\n");
            return 0;
        }
        else if (length == 16 && (int)(credit / pow(10, length - 2)) > 50 && (int)(credit / pow(10, length - 2)) < 56)
        {
            printf("MASTERCARD\n");
            return 0;
        }
        else if(length == 15 && ((int)(credit / pow(10, length - 2)) == 34 || (int)(credit / pow(10, length - 2)) == 37))
        {
            printf("AMEX\n");
            return 0;
        }
    }

    printf("INVALID\n");
}

int get_length(long num)
{
    int count = 1;
    while (num / 10 != 0)
    {
        count++;
        num = num / 10;
    }
    return count;
}

bool luhn(int length, long credit)
 {
    int sum = 0;
    bool is_odd = 1;
    while(credit != 0)
    {
        int digit = credit % 10;
        credit = credit / 10;

        if (is_odd == true)
        {
            sum += digit;
            is_odd = false;
        }
        else
        {
            if (digit > 4)
            {
                sum = sum + digit * 2 - 9;
            }
            else
            {
                sum += digit * 2;
            }
            is_odd = true;
        }
    }
    return !(sum % 10);
 }

File: week_6/sentimental-credit/credit.py
import re
def main():
    user_input = input("Number: ")

    length = len(user_input)

    if(through_luhn(user_input)):
        if (length == 13 or length == 16) and user_input.startswith('4'):
            print("VISA")
            return
        elif length == 16 and re.search('^5[1-5][0-9]*$', user_input):
            print("MASTERCARD")
            return
        elif length == 15 and (user_input.startswith("34") or user_input.startswith("37")):
            print("AMEX")
            return
    print("INVALID")

def through_luhn(n):
    sum = 0
    length = len(n)
    for i in range(length - 2, -1, -2):
        digit = int(n[i])
        if (digit > 4):
            sum += (digit * 2 - 9)
        else:
            sum += (2 * digit)

    for j in range(length - 1, -1, -2):
        sum += int(n[j])
    return not (sum % 10)


main()

